src/gridfs/download.rs

```diff
@@ -139,16 +142,206 @@ impl GridFsBucket {
+    // unwrapped safely.
+    Idle(Option<Idle>),
+    Busy(GetBytesFuture),
+    Done,
+}
+
+struct Idle {
+    buffer: Vec<u8>,
+    cursor: Box<Cursor<Chunk<'static>>>,
+}
+
+impl State {
+    fn set_busy(&mut self, new_future: GetBytesFuture) -> &mut GetBytesFuture {
+        *self = State::Busy(new_future);
+        match self {
+            Self::Busy(ref mut future) => future,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl GridFsDownloadStream {
+    async fn new(
+        file: FilesCollectionDocument,
+        chunks: &Collection<Chunk<'static>>,
+    ) -> Result<Self> {
+        let initial_state = if file.length == 0 {
+            State::Done
+        } else {
+            let options = FindOptions::builder().sort(doc! { "n": 1 }).build();
+            let cursor = chunks.find(doc! { "files_id": &file.id }, options).await?;
+            State::Idle(Some(Idle {
+                buffer: Vec::new(),
+                cursor: Box::new(cursor),
+            }))
+        };
+        Ok(Self {
+            state: initial_state,
+            current_n: 0,
+            file,
+        })
+    }
+
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.file.id
+    }
+}
+
+impl AsyncRead for GridFsDownloadStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::result::Result<usize, futures_util::io::Error>> {
+        let stream = self.get_mut();
+
+        let result = match &mut stream.state {
+            State::Idle(idle) => {
+                let Idle { buffer, cursor } = idle.take().unwrap();
+
+                if !buffer.is_empty() {
+                    Ok((buffer, cursor))
+                } else {
+                    let chunks_in_buf = FilesCollectionDocument::n_from_vals(
+                        buf.len() as u64,
+                        stream.file.chunk_size,
+                    );
+                    // We should read from current_n to chunks_in_buf + current_n, or, if that would
+                    // exceed the total number of chunks in the file, to the last chunk in the file.
+                    let final_n = std::cmp::min(chunks_in_buf + stream.current_n, stream.file.n());
+                    let n_range = stream.current_n..final_n;
+
+                    stream.current_n = final_n;
+
+                    let new_future = stream.state.set_busy(
+                        get_bytes(
+                            cursor,
+                            buffer,
+                            n_range,
+                            stream.file.chunk_size,
+                            stream.file.length,
+                        )
+                        .boxed(),
+                    );
+
+                    match new_future.poll_unpin(cx) {
+                        Poll::Ready(result) => result,
+                        Poll::Pending => return Poll::Pending,
```

Comment:
reading over the docs for `Poll::Pending` I see:

> When a function returns Pending, the function must also ensure that the current task is scheduled to be awoken when progress can be made.

Is that requirement covered somehow here?

Reply:
Hm, the reference implementations I looked at when writing this don't do anything with the waker. I believe this is a lower-level detail handled by the future we pass `cx` into -- this function itself wouldn't be able to know when that future is ready to make progress.

Reply:
👍 cool, yeah I wasn't sure how we would be expected to know when to wake either. sgtm